### PR TITLE
Add automation scripts for workflow and linter templates

### PR DESCRIPTION
**Pull request by jordanjones243**
# Alaska Airlines Pull Request

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

**Fixes:** #32, #35, #36

## Summary:

_Please summarize the scope of the changes you have submitted, what the intent of the work is and anything that describes the before/after state of the project._

* Add automation scripts that will automatically update templates for Github workflows and Linter configurations
* Add workflow that will create a surge demo for a component when a PR is created to merge into `main` or `master`
* Add automation to replace instances of `abstractNodeVersion` and `branchName` variables with a repo's respective node version and head branch name (`main` or `master`)

## Additional Context

These scripts have been tested and validated as a POC in `auro-combobox` and can be verified by @jason-capsule42.

## Type of change:

Please delete options that are not relevant.

- [x] New capability
- [x] Revision of an existing capability
- [ ] Infrastructure change (automation, etc.)
- [ ] Other (please elaborate)

## Checklist:

- [x] My update follows the CONTRIBUTING guidelines of this project
- [x] I have performed a self-review of my own update

**By submitting this Pull Request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

_Pull Requests will be evaluated by their quality of update and whether it is consistent with the goals and values of this project. Any submission is to be considered a conversation between the submitter and the maintainers of this project and may require changes to your submission._

**Thank you for your submission!**<br>
-- Auro Design System Team

### CODE REVIEW

File: scripts/config/syncWorkflows.mjs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env node
+
+// Copyright (c) Alaska Air. All right reserved. Licensed under the Apache-2.0 license
+// See LICENSE in the project root for license information.
+
+// ---------------------------------------------------------------------
+
```
Comment (jason-capsule42) [inline scripts/config/syncWorkflows.mjs:7]: Let's add some comment text in here explaining what running this script will do in plain english.

```js
/** 
* This script is for the purpose of keeping a component repository in sync with the latest set of workflows defined for Auro components. 
* Running this script will delete all workflows that exist in the repository and replace them with the latest defined workflows for Auro components. **/
```

No need to use that exact language but something like that. 

Let's add a descriptive section like this to the top of all three files
1. syncAllTemplates
1. syncLinters
1. syncWorkflows

**Review by jason-capsule42 — CHANGES_REQUESTED** [review]

File: scripts/utils/auroLibraryUtils.mjs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+      const stat = fs.statSync(sourceFilePath);
+  
+      if (stat.isDirectory()) {
+        this.copyDirectory(srcDir, destDir, removeFiles);
+      } else {
+        try {
+          fs.copyFileSync(sourceFilePath, destFilePath);
+          this.auroLogger(`Copied file: ${file}`, 'success');
+        } catch (err) {
+          this.auroLogger(`Error copying file ${file}: ${err}`, 'error');
```
[the remark this answers is not in the document]
Comment (jordanjones243) [reply scripts/utils/auroLibraryUtils.mjs:70]: No I was not able to test the logger. The way that the code is written is that if there is no existing destination directory, it creates that directory as well.

File: scripts/config/syncWorkflows.mjs (hunk 1 above)
Comment (jordanjones243) [reply scripts/config/syncWorkflows.mjs:7]: Resolved.

File: scripts/config/syncAllTemplates.mjs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+// Copyright (c) Alaska Air. All right reserved. Licensed under the Apache-2.0 license
+// See LICENSE in the project root for license information.
+
+// ---------------------------------------------------------------------
+
+/** 
+* This script is for the purpose of keeping a component repository in sync with the latest set of workflows defined for Auro components. 
+* Running this script will delete all workflows that exist in the repository and replace them with the latest defined workflows for Auro components. **/
+
+import AuroWorkflows from './syncWorkflows.mjs';
```
Comment (blackfalcon) [inline scripts/config/syncAllTemplates.mjs:12]: None of these files will be in the npm package due to this 

https://github.com/AlaskaAirlines/auro-library/blob/9a6a7d05783fd1f3be5cd06e75b2bf57effab83f/.npmignore#L2

File: componentTemplates/linters/.eslintrc [hunk 4]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (blackfalcon) [inline componentTemplates/linters/.eslintrc:1]: We need a story to remove these from the Generator and then the Generator will require a dependency on this package with a script to install its contents.

File: componentTemplates/workflows/codeql.yml [hunk 5]
```diff
@@ -0,0 +1,58 @@
+name: "CodeQL"
+
+on:
+  push:
+    branches: [ "master" ]
+  pull_request:
+    # The branches below must be a subset of the branches above
+    branches: [ "master" ]
```
Comment (blackfalcon) [inline componentTemplates/workflows/codeql.yml:8]: This needs to be configurable based on what kind of repo it's being added to. I would use `main` as the default with an option to change to `master`.

File: componentTemplates/workflows/testPublish.yml [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    branches: [ master]
+
+# A workflow run is made up of one or more jobs that can run sequentially or in parallel
+jobs:
+  test:
+    runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        node-version: [18.x]
```
Comment (blackfalcon) [inline componentTemplates/workflows/testPublish.yml:18]: Again, this is not static. In the generator, we use an app variable that is set to whatever the currently supported version is. 

https://github.com/AlaskaAirlines/WC-Generator/blob/3a9b8996afb6509e92d58869ad024e33bbc5a813/package.json#L20-L22

https://github.com/AlaskaAirlines/WC-Generator/blob/3a9b8996afb6509e92d58869ad024e33bbc5a813/bin/generate.js#L259

File: componentTemplates/workflows/testPublish.yml [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    if: github.event_name == 'push' && github.ref == 'refs/heads/master'
+    runs-on: ubuntu-latest
+    needs: test
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          persist-credentials: false
+      - uses: actions/setup-node@v3
+        with:
+          node-version: 18
```
Comment (blackfalcon) [inline componentTemplates/workflows/testPublish.yml:40]: Same as above

**Review by blackfalcon — CHANGES_REQUESTED** [review]
So, the intent is to put things into this repo/npm and remove them from the Generator? Also, what is the process for running these scripts?

File: scripts/utils/auroLibraryUtils.mjs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jason-capsule42) [reply scripts/utils/auroLibraryUtils.mjs:70]: What about giving it a bad source directory/file path?

[the remark this answers is not in the document]
Comment (jordanjones243) [reply scripts/utils/auroLibraryUtils.mjs:70]: When giving a bad source directory path, an error is thrown in the console, but not from the `auroLogger`. Here is an example output of what comes back from an incorrect file path.
Screen Shot 2023-08-23 at 11 27 35 AM

**Comment by jordanjones243** [comment]
> So, the intent is to put things into this repo/npm and remove them from the Generator? Also, what is the process for running these scripts?

Yes, the intent is to remove workflow and configuration files from the Generator and move them into `auro-library`. A story for this can be created by @braven112 or @Patrick-Daly-AA.

The process for running/testing these scripts is to link `auro-library` into another component. To run the `syncAllWorkflows.mjs` script, you will need to add a new node script into the linked component and point that to the `syncAllWorkflows.mjs` file. You can individually run the workflow configurations by pointing to the `syncAllWorkflows.mjs` file and adding a `--github` parameter after the path. The same can be done for the linter configurations by adding a `--linters` parameter.

To run the updated `generateDocs.mjs` script, point the `build:docs` script in the linked component to the the one that is now in `auro-library`, rather than the one currently being used in the linked component's repo.

File: scripts/utils/auroLibraryUtils.mjs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jordanjones243) [reply scripts/utils/auroLibraryUtils.mjs:75]: To answer your question, I do not believe the `catch` branch is ever reached because the error shown in the screenshot is thrown before the catch function is ever reached.

**Comment by blackfalcon** [comment]
@jordanjones243 

1. Engineers are responsible for creating issues with the necessary details for the work. @Patrick-Daly-AA and @braven112 really have the decision on when to do the work. 
2. All the instructions you put in this comment, this should be put into the README or other resource. This isn't directly intuitive and no one is going to dig into these comments for these instructions.

**Comment by jordanjones243** [comment]
> @jordanjones243
> 
> 1. Engineers are responsible for creating issues with the necessary details for the work. @Patrick-Daly-AA and @braven112 really have the decision on when to do the work.
> 2. All the instructions you put in this comment, this should be put into the README or other resource. This isn't directly intuitive and no one is going to dig into these comments for these instructions.

I have created a story in `WC-Generator` to move the templates from the generator to `auro-library`:
* https://github.com/AlaskaAirlines/WC-Generator/issues/423

I have also created a `.md` file containing instructions and code examples of how to run these scripts.

File: docs/syncAllTemplates.md [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+```
+// Only sync github workflow templates
+"syncTemplates": "./node_modules/@aurodesignsystem/auro-library/scripts/setup/syncAllTemplates.mjs --github"
+```
+
+```
+// Only sync linter configuration templates
+"syncTemplates": "./node_modules/@aurodesignsystem/auro-library/scripts/setup/syncAllTemplates.mjs --linters"
+```
```
Comment (blackfalcon) [inline docs/syncAllTemplates.md:20]: Please either have this in the README or there is a reference to this document in the README. The `./docs` directory is mainly used for secondary information or long descriptions of repo documentation. 

https://en.wikipedia.org/wiki/README

A README file typically encompasses:
- Configuration instructions
- Installation instructions
- **Operating instructions**
- A file manifest (a list of files in the directory or archive)
- Copyright and licensing information
- Contact information for the distributor or author
- A list of known bugs[3]
- Troubleshooting instructions[3]
- Credits and acknowledgments
- A changelog (usually aimed at fellow programmers)
- A news section (usually aimed at end users)

**Review by blackfalcon — CHANGES_REQUESTED** [review]

**Comment by jason-capsule42** [comment]
> > @jordanjones243
> > 
> > 1. Engineers are responsible for creating issues with the necessary details for the work. @Patrick-Daly-AA and @braven112 really have the decision on when to do the work.
> > 2. All the instructions you put in this comment, this should be put into the README or other resource. This isn't directly intuitive and no one is going to dig into these comments for these instructions.
> 
> I have created a story in `WC-Generator` to move the templates from the generator to `auro-library`:
> 
> * [Move templates over into auro-library WC-Generator#423](https://github.com/AlaskaAirlines/WC-Generator/issues/423)
> 
> I have also created a `.md` file containing instructions and code examples of how to run these scripts.

@blackfalcon @jordanjones243 I created a ticket that we should get prioritized soon for getting the individual documentation files rolled up into a more thorough readme and presented onto the docsite.

https://github.com/AlaskaAirlines/auro-library/issues/37

**Review by jason-capsule42 — APPROVED** [review]
